### PR TITLE
Adding option to monitor other docker containers

### PR DESCRIPTION
**Pull request by c0derMo**
# Description

Adds the ability to monitor the uptime of other docker containers on the same system.
Docker daemon & container are specified for each monitor.
Axios is used to call the docker API at the specified daemon directly, inspecting the target container to figure out if its running.

## Type of change

- New feature (non-breaking change which adds functionality)

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I ran ESLint and other linters for modified files
- [X] I have performed a self-review of my own code and tested it
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

### CODE REVIEW

File: server/database.js [hunk 1]
```diff
@@ -53,6 +53,7 @@ class Database {
         "patch-2fa-invalidate-used-token.sql": true,
         "patch-notification_sent_history.sql": true,
         "patch-monitor-basic-auth.sql": true,
+        "patch-add-docker-columns.sql": true
```
Comment (Saibamen) [inline server/database.js:56]:
```suggestion
        "patch-add-docker-columns.sql": true,
```

File: server/model/monitor.js [hunk 2]
```diff
@@ -347,6 +349,28 @@ class Monitor extends BeanModel {
                         throw new Error("Server not found on Steam");
                     }
 
+                } else if (this.type === "docker") {
+                    debug(`[${this.name}] Prepare Options for axios`);
```
Comment (Saibamen) [inline server/model/monitor.js:353]:
```suggestion
                    debug(`[${this.name}] Prepare Options for Axios`);
```

File: src/pages/EditMonitor.vue [hunk 3]
```diff
@@ -32,6 +32,9 @@
                                     <option value="steam">
                                         Steam Game Server
                                     </option>
+                                    <option value="docker">
+                                        Docker Container
```
Comment (Saibamen) [inline src/pages/EditMonitor.vue:36]: Not translated

File: src/pages/EditMonitor.vue [hunk 4]
```diff
@@ -439,6 +456,8 @@ export default {
                     accepted_statuscodes: ["200-299"],
                     dns_resolve_type: "A",
                     dns_resolve_server: "1.1.1.1",
+                    docker_container: "",
+                    docker_daemon: "/var/run/docker.sock"
```
Comment (Saibamen) [inline src/pages/EditMonitor.vue:460]:
```suggestion
                    docker_daemon: "/var/run/docker.sock",
```

**Review by Saibamen — CHANGES_REQUESTED** [review]

File: src/pages/EditMonitor.vue (hunk 3 above)
Comment (c0derMo) [reply src/pages/EditMonitor.vue]: Added the translation, just asking though, why does this get translated, but e.g. "Steam Game Server" doesn't?

Comment (Saibamen) [reply src/pages/EditMonitor.vue]: Some PR authors forget about it and this wasn't catch in Code Review

**Comment by c43721** [comment]
this is kinda neat, right now all I do to ping my containers is do a TCP ping on the container name, since my containers are all on the same network. This I think means that I don't need to do that anymore!!

I thank you for this contribution.

**Comment by chakflying** [comment]
Is there any way to make this work on Windows? I did a quick search with no success.

**Comment by c0derMo** [comment]
@chakflying 
> Is there any way to make this work on Windows? I did a quick search with no success.

Doesn't Docker for Windows just publish the daemon using tcp?
(Can't verify rn since I'm not on my windows machine, but looking at the docs I would assume it working that way)
Because if that's the case, I can just add a menu to toggle between unix-socket and tcp for connecting to the daemon

**Comment by c0derMo** [comment]
With this commit, the uptime monitor is able to connect to the docker daemon via tcp / http, which allows for windows compatiblity.
If the setting "Expose daemon on tcp://localhost:2375" is enabled in the Docker for Windows settings, the uptime monitor can connect to it using TCP or HTTP, and it will work just like with a unix socket.

**Comment by c0derMo** [comment]
Any updates on this? It's been a while, and I'd still love to see this feature implemented

File: server/model/monitor.js [hunk 5]
```diff
@@ -347,6 +350,33 @@ class Monitor extends BeanModel {
...
+                    } else if (this.docker_type === "tcp") {
+                        options.baseURL = this.docker_daemon;
+                    }
+
+                    debug(`[${this.name}] Axios Request`);
+                    let res = await axios.request(options);
+                    if (res.data.State.Running) {
+                        bean.status = UP;
+                        bean.msg = "";
+                    }
```
Comment (Computroniks) [inline server/model/monitor.js:379]:
```suggestion
                    }
                    
```

Perhaps add a new line here to match up with before and the rest of the `if else` block.

**Review by Computroniks — CHANGES_REQUESTED** [review]

**Review by Computroniks — APPROVED** [review]

**Comment by joerich** [comment]
some insight into how to invoke this little gem would be awesome!

**Comment by chrisbecke** [comment]
I was thinking of a similar thing, but for docker swarm:
I want a health check that counts the number of (healthy) running replicas for a service compared to the number of requested replicas.

**Comment by c0derMo** [comment]
Let's first work on getting this PR merged, before adding even more features that need to be audited.

**Comment by louislam** [comment]
I think maybe we should make `docker_daemon` and `docker_type` into a table. Just similar to notification, user can add/edit/test the docker connection in a setting page.

For future development, we can change the container name text field into dropdown menu. The UX will be better.

**Comment by c0derMo** [comment]
Took me a bit, but I've added a seperate table for docker hosts, as per @louislam 's request.
It's not 100% finished yet, some localization & a settings page is missing, that'll get done in the near future.
Any feedback / review is appreciated, I'm not too familiar with the codebase yet, looked at proxies & notifications for reference.

**Comment by c0derMo** [comment]
From my point of view, this is ready for merging now, with the changes louislam suggested. Any reviews or feedback is still greatly appreciated.

**Comment by louislam** [comment]
I found that the test button and save button both were not working.

I have fixed it, feel free to test it.

**Pull request merged**